streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
                        stateUpdateSupplier);
     }
 
+    public void buildAndOptimizeTopology(Properties props) {
+        if (!topologyBuilt) {
+
+            if (props != null && props.getProperty(StreamsConfig.TOPOLOGY_OPTIMIZATION).equals(StreamsConfig.OPTIMIZE)) {
+                LOG.debug("Optimizing the Kafka Streams graph");
+                optimize();
+            }
+            final PriorityQueue<StreamsGraphNode> graphNodePriorityQueue = new PriorityQueue<>(5, nodeIdComparator);
+
+            graphNodePriorityQueue.offer(root);
+
+            while (!graphNodePriorityQueue.isEmpty()) {
+                final StreamsGraphNode streamGraphNode = graphNodePriorityQueue.remove();
+
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("about write topology for {} child nodes {}", streamGraphNode, streamGraphNode.children());
+                }
+
+                streamGraphNode.writeToTopology(internalTopologyBuilder);
+
+                for (StreamsGraphNode graphNode : streamGraphNode.children()) {
+                    graphNodePriorityQueue.offer(graphNode);
+                }
+            }
+
+            topologyBuilt = true;
+        }
+    }
+
+    private void optimize() {
+        mayOptimizeRepartitionOperations();
```

Comment:
Meta comment: if we are going to add more optimization rules in `optimize()`, should we keep this PR as a plain one that do not enforce any optimizations, so that we can then consider each function separately, that helps more concentrated reviews and reduce large PR burdens.

Reply:
Ack removed optimization in favor of pushing a 4th PR with optimizations only once 3rd PR is merged.